src/server/pfs/server/driver.go

```diff
@@ -800,188 +706,117 @@ func (d *driver) repoSize(ctx context.Context, repo *pfs.Repo) (int64, error) {
+			if _, ok := seen[pfsdb.BranchKey(sb)]; !ok {
+				sbi := &pfs.BranchInfo{}
+				if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(sb, sbi); err != nil {
+					return errors.Wrapf(err, "get subvenant branch %q", pfsdb.BranchKey(sb))
+				}
+				seen[pfsdb.BranchKey(sb)] = sbi
+				propagatedBranches = append(propagatedBranches, seen[pfsdb.BranchKey(sb)])
 			}
-			subvBIMap[pfsdb.BranchKey(subvBranch)] = subvInfo
 		}
 	}
-
-	// 'subvBIs' is the collection of downstream branches that will get a new
-	// commit. Populate subvBIs and sort it so that upstream branches are
-	// processed before their descendants (this guarantees that if branch B is
-	// provenant on branch A, we create a new commit in A before creating a new
-	// commit in B provenant on the new HEAD of A).
-	var subvBIs []*pfs.BranchInfo
-	for _, branchData := range subvBIMap {
-		subvBIs = append(subvBIs, branchData)
-	}
-	sort.Slice(subvBIs, func(i, j int) bool {
-		return len(subvBIs[i].Provenance) < len(subvBIs[j].Provenance)
+	sort.Slice(propagatedBranches, func(i, j int) bool {
+		return len(propagatedBranches[i].Provenance) < len(propagatedBranches[j].Provenance)
 	})
-
-	// Iterate through downstream branches and add new commits.
-	hasNewCommits := false
-	for _, subvBI := range subvBIs {
+	// add new commits, set their ancestry + provenance pointers, and advance branch heads
+	for _, bi := range propagatedBranches {
+		// TODO(acohen4): can we just make calls to startCommit() here?
 		// Do not propagate an open commit onto spout output branches (which should
 		// only have a single provenance on a spec commit)
-		if len(subvBI.Provenance) == 1 && subvBI.Provenance[0].Repo.Type == pfs.SpecRepoType {
+		if len(bi.Provenance) == 1 && bi.Provenance[0].Repo.Type == pfs.SpecRepoType {
 			continue
 		}
-		hasNewCommits = true
-
-		// Create aliases for any provenant branches which are not already part of this CommitSet
-		for _, provOfSubvB := range subvBI.Provenance {
-			provOfSubvBI, err := getBranchInfo(provOfSubvB)
-			if err != nil {
-				return err
-			}
-			if provOfSubvBI.Head.ID != txnCtx.CommitSetID {
-				if _, err := d.aliasCommit(txnCtx, provOfSubvBI.Head, provOfSubvBI.Head.Branch); err != nil {
-					return err
-				}
-				// Update the cached branch head
-				provOfSubvBI.Head.ID = txnCtx.CommitSetID
-			}
-			// if this is a pipeline output branch, we need to also create an alias commit on the meta branch
-			// to maintain pipeline system invariants.
-			if provOfSubvBI.Branch.Repo.Type == pfs.UserRepoType {
-				metaBranch := client.NewSystemProjectRepo(provOfSubvBI.Branch.Repo.Project.GetName(), provOfSubvBI.Branch.Repo.Name, pfs.MetaRepoType).
-					NewBranch(provOfSubvBI.Branch.Name)
-				metaBI, err := getBranchInfo(metaBranch)
-				if err != nil {
-					if col.IsErrNotFound(err) {
-						// no corresponding meta branch, so not a pipeline. Ignore
-						continue
-					}
-					return err
+		// TODO(acohen4): understand why this is necessary
```

Comment:
Undone TODO.

Reply:
This might have been necessary in the squash / drop commit case since they can / could create commits outside of propagation.